airbyte-workers/src/main/java/io/airbyte/workers/process/ExitCodeWatcher.java

```diff
@@ -20,52 +20,101 @@
  * be able to retrieve the exit code).
  */
 @Slf4j
-public class ExitCodeWatcher implements Watcher<Pod> {
+public class ExitCodeWatcher implements ResourceEventHandler<Pod> {
 
+  private final String podName;
+  private final String podNamespace;
   private final Consumer<Integer> onExitCode;
-  private final Consumer<WatcherException> onWatchFailure;
-  private boolean exitCodeRetrieved = false;
+  private final Runnable onWatchFailure;
+  /**
+   * This flag is set to false when we either (a) find the pod's exit code, or (b) when the pod is
+   * deleted. This is so that we call exactly one of onExitCode and onWatchFailure, and we make that
+   * call exactly once.
+   */
+  private boolean active = true;
 
   /**
-   *
    * @param onExitCode callback used to store the exit code
    * @param onWatchFailure callback that's triggered when the watch fails. should be some failed exit
    *        code.
    */
-  public ExitCodeWatcher(final Consumer<Integer> onExitCode, final Consumer<WatcherException> onWatchFailure) {
+  public ExitCodeWatcher(final String podName,
+                         final String podNamespace,
+                         final Consumer<Integer> onExitCode,
+                         final Runnable onWatchFailure) {
```

Comment:
why are we exposing this? is this for testing or something, if so we can we put it in a non-public constructor? looks like we pass in a no op here, no?

Reply:
Similar thing as above, the production case is logging a message + setting the exit code future to 143 https://github.com/airbytehq/airbyte/pull/11861/files#diff-be3c5f5544310ca55f284848493a7afe72533bd82a737a044d3f58e9c82d71abR540-R546

This used to accept an Exception, but since Informers handle errors transparently for us, this is basically just a failsafe for the pod somehow not triggering the watch but still getting deleted. I'll add a comment saying that this should be impossible / very unlikely.